components/brave_wallet/browser/siwe_message_parser_unittest.cc

```diff
@@ -0,0 +1,497 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <string>
+
+#include "base/strings/strcat.h"
+#include "brave/components/brave_wallet/browser/siwe_message_parser.h"
+#include "brave/components/brave_wallet/common/sign_message_request.mojom.h"
+#include "testing/gtest/include/gtest/gtest.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+namespace brave_wallet {
+
+namespace {
+constexpr char kExampleAddress[] = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2";
+constexpr char kExampleUri[] = "https://example.com/login";
+constexpr char kExampleNonce[] = "32891756";
+constexpr char kExampleISO8601Time[] = "2021-09-30T16:25:24Z";
+constexpr char kExampleRequestId[] = "8b270b8a-d74b-459b-9933-81cb234d7c5e";
+constexpr char kExampleResourceURI1[] =
+    "ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq";
+constexpr char kExampleResourceURI2[] =
+    "https://example.com/my-web2-claim.json";
+}  // namespace
+
+using State = SIWEMessageParser::State;
+
+class SIWEMessageParserTest : public testing::Test {
```

Comment:
nit: for robustness could we copy the [parsing cases from the reference implementation](https://github.com/spruceid/siwe/tree/main/test) and reuse them as unit tests here?

Reply:
Our unit test is pretty robust enough and we don't need to copy anything. And please note stuffs that wallet verify are different than other roles, ex. wallet doesn't verify if those time field expired or not.
So we should not blindly copy tests from others. If we missed anything, we can add anytime.

Reply:
MetaMask is directly relying on the reference implementation hence the request. So by utilizing these tests we can maintain a higher assurance of compatibility with Metamask and the reference implementation at the unit tests. It's not to say that what we've implemented is wrong, in fact it's quite comprehensive. Instead it's about maintaining compatibility with other implementations.

As you pointed out we're not checking expiration time, but this seems like a case where it would be useful to notify the user of this issue via the warning panel. 

As we perform refactors over this code or add new signing functionality it seems useful that we'd be able to detect any regressions early. The alternative approach I could see us taking is to integrate compatibility tests like this via the wallet test framework (later) and ignore addressing compatibility aspects for now. Wdyt about that option?

Reply:
Those time fields are not required to be verify by wallet on the spec and wallet doesn't use any of those fields and if even if needs to verify that we need to make sure users clock is at least in sync.

You can open an issue for follow-up integrating those tests.

Reply:
I agree with you in terms of strictness of the rules. We're adhering to the spec properly too.  I can think of a few good reasons for and against implementing it, but I'm curious what your reasoning was for not implementing this?

Also, thank you for calling this case out so we can discuss it a bit further and make sure we're all aligned on what we should do. I didn't catch this previously or else I would have raised it for us to discuss further.

As for clock syncing issue what I've normally seen within OAuth and other JWT based token implementations is that the verifier will accept a certain level of clock skew (say 10 seconds) in order to handle this as a "good enough" solution. We could perhaps address it in a similar way if we decide to go forward with it.

Reply:
It's very simple, those timestamps hold no meaning to clients we are not using it and I don't think it worth the effort implementing it just because we can.

Reply:
Ok, I just realized that I don't think I'm aligned on that view, but this doesn't need to block it. What you've implemented is correct. I'll kick off a separate discussion in slack and if we want to change it do it in a follow up PR.